File: scrapers/souped_up/yahoo_financial_statements_scraper.py
import pandas as pd
import urllib.request as ur
from bs4 import BeautifulSoup
import warnings

def yahoo_income_statement(ticker='AAPL'):
    income_url = f'https://finance.yahoo.com/quote/{ticker}/financials?p={ticker}'
    read_url = ur.urlopen(income_url).read()
    income_soup = BeautifulSoup(read_url, 'lxml')
    div_list = []
    for div in income_soup.find_all('div'):
        div_list.append(div.string)
        if not div.string == div.get('title'):
            div_list.append(div.get('title'))
    div_list = [incl for incl in div_list if incl not in ('Operating Expenses', 'Non-recurring Events', 'Expand All')]
    div_list = list(filter(None, div_list))
    div_list = [incl for incl in div_list if not incl.startswith('(function')]
    income_list = div_list[13: -5]
    income_list.insert(0, 'Breakdown')
    income_data = list(zip(*[iter(income_list)]*6))
    income_df = pd.DataFrame(income_data)
    headers = income_df.iloc[0]
    income_df = income_df[1:]
    income_df.columns = headers
    income_df.set_index('Breakdown', inplace=True, drop=True)
    warnings.warn('Amounts are in thousands.')
    return income_df

def yahoo_balance_sheet(ticker='AAPL'):
    balancesheet_url = f'https://finance.yahoo.com/quote/{ticker}/balance-sheet?p={ticker}'
    read_url = ur.urlopen(balancesheet_url).read()
    balancesheet_soup = BeautifulSoup(read_url, 'lxml')
    div_list = []
    for div in balancesheet_soup.find_all('div'):
        div_list.append(div.string)
        if not div.string == div.get('title'):
            div_list.append(div.get('title'))
    div_list = [incl for incl in div_list if incl not in ('Operating Expenses', 'Non-recurring Events', 'Expand All')]
    div_list = list(filter(None, div_list))
    div_list = [incl for incl in div_list if not incl.startswith('(function')]
    balancesheet_list = div_list[13: -5]
    balancesheet_list.insert(0, 'Breakdown')
    balancesheet_data = list(zip(*[iter(balancesheet_list)]*5))
    balancesheet_df = pd.DataFrame(balancesheet_data)
    headers = balancesheet_df.iloc[0]
    balancesheet_df = balancesheet_df[1:]
    balancesheet_df.columns = headers
    balancesheet_df.set_index('Breakdown', inplace=True, drop=True)
    warnings.warn('Amounts are in thousands.')
    return balancesheet_df

def yahoo_cash_flow(ticker='AAPL'):
    cashflow_url = f'https://finance.yahoo.com/quote/{ticker}/cash-flow?p={ticker}'
    read_url = ur.urlopen(cashflow_url).read()
    cashflow_soup = BeautifulSoup(read_url, 'lxml')
    div_list = []
    for div in cashflow_soup.find_all('div'):
        div_list.append(div.string)
        if not div.string == div.get('title'):
            div_list.append(div.get('title'))
    div_list = [incl for incl in div_list if incl not in ('Operating Expenses', 'Non-recurring Events', 'Expand All')]
    div_list = list(filter(None, div_list))
    div_list = [incl for incl in div_list if not incl.startswith('(function')]
    cashflow_list = div_list[13: -5]
    cashflow_list.insert(0, 'Breakdown')
    cashflow_data = list(zip(*[iter(cashflow_list)]*6))
    cashflow_df = pd.DataFrame(cashflow_data)
    headers = cashflow_df.iloc[0]
    cashflow_df = cashflow_df[1:]
    cashflow_df.columns = headers
    cashflow_df.set_index('Breakdown', inplace=True, drop=True)
    warnings.warn('Amounts are in thousands.')
    return cashflow_df

File: my-app/lib/scrapers/yahoo_financial_statements_scraper.js
const axios = require('axios');
const cheerio = require('cheerio');

async function yahooIncomeStatement(ticker = 'AAPL') {
  const incomeUrl = `https://finance.yahoo.com/quote/${ticker}/financials?p=${ticker}`;
  try {
    const { data } = await axios.get(incomeUrl);
    const $ = cheerio.load(data);
    let divList = [];
    $('div').each((i, div) => {
      const text = $(div).text().trim();
      if (text) {
        divList.push(text);
      }
    });
    divList = divList.filter(item => !['Operating Expenses', 'Non-recurring Events', 'Expand All'].includes(item));
    divList = divList.filter(item => item && !item.startsWith('(function'));
    let incomeList = divList.slice(13, -5);
    incomeList.unshift('Breakdown');
    const incomeData = [];
    for (let i = 0; i < incomeList.length; i += 6) {
      incomeData.push(incomeList.slice(i, i + 6));
    }
    const headers = incomeData[0];
    incomeData.shift();
    const incomeObj = incomeData.map(row => {
      let result = {};
      headers.forEach((header, i) => {
        result[header] = row[i];
      });
      return result;
    });
    console.warn('Amounts are in thousands.');
    return incomeObj;
  } catch (error) {
    console.error('Error fetching income statement:', error);
  }
}

async function yahooBalanceSheet(ticker = 'AAPL') {
  const balanceSheetUrl = `https://finance.yahoo.com/quote/${ticker}/balance-sheet?p=${ticker}`;
  try {
    const { data } = await axios.get(balanceSheetUrl);
    const $ = cheerio.load(data);
    let divList = [];
    $('div').each((i, div) => {
      const text = $(div).text().trim();
      if (text) {
        divList.push(text);
      }
    });
    divList = divList.filter(item => !['Operating Expenses', 'Non-recurring Events', 'Expand All'].includes(item));
    divList = divList.filter(item => item && !item.startsWith('(function'));
    let balanceSheetList = divList.slice(13, -5);
    balanceSheetList.unshift('Breakdown');
    const balanceSheetData = [];
    for (let i = 0; i < balanceSheetList.length; i += 5) {
      balanceSheetData.push(balanceSheetList.slice(i, i + 5));
    }
    const headers = balanceSheetData[0];
    balanceSheetData.shift();
    const balanceSheetObj = balanceSheetData.map(row => {
      let result = {};
      headers.forEach((header, i) => {
        result[header] = row[i];
      });
      return result;
    });
    console.warn('Amounts are in thousands.');
    return balanceSheetObj;
  } catch (error) {
    console.error('Error fetching balance sheet:', error);
  }
}

async function yahooCashFlow(ticker = 'AAPL') {
  const cashFlowUrl = `https://finance.yahoo.com/quote/${ticker}/cash-flow?p=${ticker}`;
  try {
    const { data } = await axios.get(cashFlowUrl);
    const $ = cheerio.load(data);
    let divList = [];
    $('div').each((i, div) => {
      const text = $(div).text().trim();
      if (text) {
        divList.push(text);
      }
    });
    divList = divList.filter(item => !['Operating Expenses', 'Non-recurring Events', 'Expand All'].includes(item));
    divList = divList.filter(item => item && !item.startsWith('(function'));
    let cashFlowList = divList.slice(13, -5);
    cashFlowList.unshift('Breakdown');
    const cashFlowData = [];
    for (let i = 0; i < cashFlowList.length; i += 6) {
      cashFlowData.push(cashFlowList.slice(i, i + 6));
    }
    const headers = cashFlowData[0];
    cashFlowData.shift();
    const cashFlowObj = cashFlowData.map(row => {
      let result = {};
      headers.forEach((header, i) => {
        result[header] = row[i];
      });
      return result;
    });
    console.warn('Amounts are in thousands.');
    return cashFlowObj;
  } catch (error) {
    console.error('Error fetching cash flow:', error);
  }
}

// yahooIncomeStatement('AAPL').then(data => console.log(data));
// yahooBalanceSheet('AAPL').then(data => console.log(data));
// yahooCashFlow('AAPL').then(data => console.log(data));